In [1]:
from google_play_scraper import app

In [2]:
Telegram_description = app("org.telegram.messenger" , lang = "en" , country='us')
print(Telegram_description)

{'title': 'Telegram', 'description': "Pure instant messaging — simple, fast, secure, and synced across all your devices. One of the world's top 10 most downloaded apps with over 500 million active users.\r\n\r\nFAST: Telegram is the fastest messaging app on the market, connecting people via a unique, distributed network of data centers around the globe.\r\n\r\nSYNCED: You can access your messages from all your phones, tablets and computers at once. Telegram apps are standalone, so you don’t need to keep your phone connected. Start typing on one device and finish the message from another. Never lose your data again.\r\n\r\nUNLIMITED: You can send media and files, without any limits on their type and size. Your entire chat history will require no disk space on your device, and will be securely stored in the Telegram cloud for as long as you need it. \r\n\r\nSECURE: We made it our mission to provide the best security combined with ease of use. Everything on Telegram, including chats, grou

In [3]:
from google_play_scraper import Sort, reviews

In [4]:
Telegram_reviews , continuation_token = reviews("org.telegram.messenger",lang = "en" , country='us', count = 20000)

In [5]:
Telegram_reviews = reviews('org.telegram.messenger',continuation_token=continuation_token , country='us',count = 20000 )
print(Telegram_reviews)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [6]:
Telegram_reviews_list=list(Telegram_reviews)

Telegram_reviews_list

[[{'reviewId': 'gp:AOqpTOGMZ5mdCJUFkTD3BIt0taWUQAcUL9Gvgn7cFUoAilbAD1joSsrqmeGPthqz57N8qZzGUd4WD_98OxPcAlQ',
   'userName': 'kru m',
   'userImage': 'https://play-lh.googleusercontent.com/a/AATXAJzCyMPOfisjXG_LqDRnS8qAytRR1jroWjBpZi2v=mo',
   'content': "Updated 3 days before but badly unable to send 900 mb video as tried 7 times today full day but after 920 mb it's stuck n cancel automatically....than have to tried again n it's start from zero...full day it used 6 GB to send one video but still it's failed ....if any error than why it's not start from same point",
   'score': 2,
   'thumbsUpCount': 1,
   'reviewCreatedVersion': '7.8.0',
   'at': datetime.datetime(2021, 7, 5, 0, 30, 54),
   'replyContent': None,
   'repliedAt': None},
  {'reviewId': 'gp:AOqpTOExlb93e-OzqpVB7SG_oLNrKn7ifb3npNnPAMjCn6q9oMNDpQVpMqZ1yGdS85AoPKoq-K4TpS6y8Qi3otA',
   'userName': 'Mahnoor Shahzad',
   'userImage': 'https://play-lh.googleusercontent.com/a/AATXAJzaNAUs8GClWlM5CUxkLe6qe27Wvbi92JlOrtwX=mo',
   'c

In [7]:
username = []
review=[]
user_score = []
thumbsup = []
date=[]

for x in range(0,20000):
    
    user_name=Telegram_reviews_list[0][x]["userName"]
    
    username.append(user_name)
    
    review_=Telegram_reviews_list[0][x]["content"]
    
    review.append(review_)
    
    score_=Telegram_reviews_list[0][x]["score"]
    
    user_score.append(score_)
    
    thumbsup_=Telegram_reviews_list[0][x]["thumbsUpCount"]
    
    thumbsup.append(thumbsup_)
    
    date_at=Telegram_reviews_list[0][x]["at"]
    
    date.append(date_at)



In [8]:
print(len(username))
print(len(review))
print(len(user_score))
print(len(thumbsup))
print(len(date))

20000
20000
20000
20000
20000


In [9]:
import pandas as pd

In [10]:
Telegram =pd.DataFrame({"Username":username, "Reviews": review,"Score":user_score , "ThumbsupCount" : thumbsup, "Date&Time":date})

Telegram

Username  \
0                         kru m   
1               Mahnoor Shahzad   
2                    Lucy Begam   
3                      k.k Ayan   
4             MOSKOV TOP GLOBAL   
...                         ...   
19995                Vinu Vinod   
19996  Royal Electrical company   
19997                 Ahem SaFi   
19998  Mohd7861 Salamsalimnoori   
19999               Rҽҽɱ Aʅ Fʅα   

                                                 Reviews  Score  \
0      Updated 3 days before but badly unable to send...      2   
1      It's not opening, like literally entered my nu...      1   
2                                           just awesome      5   
3                             Its soo cool and also good      5   
4      gua selalu nonton wibu , uh asik banget tegang...      5   
...                                                  ...    ...   
19995                                              Super      5   
19996                                               Good      1   
19997           This is amazing I love that application.      5   
19998                                               good      5   
19999                                               زربة      1   

       ThumbsupCount           Date&Time  
0                  1 2021-07-05 00:30:54  
1                  0 2021-07-05 00:27:56  
2                  0 2021-07-05 00:15:50  
3                  0 2021-07-05 00:13:33  
4                  0 2021-07-05 00:13:16  
...              ...                 ...  
19995              0 2021-06-08 20:59:22  
19996              0 2021-06-08 20:55:27  
19997              0 2021-06-08 20:53:40  
19998              0 2021-06-08 20:51:20  
19999              0 2021-06-08 20:49:29  

[20000 rows x 5 columns]

In [11]:
!pip install openpyxl

In [12]:
file_name = 'Telegram_20k.xlsx'
Telegram.to_excel(file_name)
print('DataFrame is written to Excel File successfully.')

DataFrame is written to Excel File successfully.


In [13]:
# Let's pick only the reviews.text column.

review_df=Telegram[['Reviews']]
review_df

Reviews
0      Updated 3 days before but badly unable to send...
1      It's not opening, like literally entered my nu...
2                                           just awesome
3                             Its soo cool and also good
4      gua selalu nonton wibu , uh asik banget tegang...
...                                                  ...
19995                                              Super
19996                                               Good
19997           This is amazing I love that application.
19998                                               good
19999                                               زربة

[20000 rows x 1 columns]

In [14]:
review_df['Reviews']=review_df['Reviews'].astype(str)

<ipython-input-14-09a84c77f960>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  review_df['Reviews']=review_df['Reviews'].astype(str)


In [15]:
review_df.isnull().sum()

Reviews    0
dtype: int64

In [16]:
review_df.dropna()

Reviews
0      Updated 3 days before but badly unable to send...
1      It's not opening, like literally entered my nu...
2                                           just awesome
3                             Its soo cool and also good
4      gua selalu nonton wibu , uh asik banget tegang...
...                                                  ...
19995                                              Super
19996                                               Good
19997           This is amazing I love that application.
19998                                               good
19999                                               زربة

[20000 rows x 1 columns]

In [17]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from sklearn.feature_extraction.text import TfidfVectorizer
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
import xgboost

In [18]:
# Create an Sentiment Intensity Analyzer object.

senti=SentimentIntensityAnalyzer()

In [19]:
# STEP 1: Create a user defined lambda function.

fun=lambda x: senti.polarity_scores(x)['compound']

In [20]:
# STEP 2: Apply the function on your data frame.

review_df['Compound']=review_df['Reviews'].apply(fun)

review_df

<ipython-input-20-496ca68a219b>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  review_df['Compound']=review_df['Reviews'].apply(fun)


Reviews  Compound
0      Updated 3 days before but badly unable to send...   -0.9528
1      It's not opening, like literally entered my nu...   -0.2047
2                                           just awesome    0.6249
3                             Its soo cool and also good    0.6369
4      gua selalu nonton wibu , uh asik banget tegang...    0.0000
...                                                  ...       ...
19995                                              Super    0.5994
19996                                               Good    0.4404
19997           This is amazing I love that application.    0.8402
19998                                               good    0.4404
19999                                               زربة    0.0000

[20000 rows x 2 columns]

In [21]:
# Let's create category labels as per the above slabs.

import numpy as np

review_df['Category']=pd.cut(review_df['Compound'],bins=[-1,-0.5,-0.2,0.2,0.5,1],labels=['High Neg','Neg','Neu','Pos','High Pos'])

review_df

<ipython-input-21-2ddee71494c7>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  review_df['Category']=pd.cut(review_df['Compound'],bins=[-1,-0.5,-0.2,0.2,0.5,1],labels=['High Neg','Neg','Neu','Pos','High Pos'])


Reviews  Compound  Category
0      Updated 3 days before but badly unable to send...   -0.9528  High Neg
1      It's not opening, like literally entered my nu...   -0.2047       Neg
2                                           just awesome    0.6249  High Pos
3                             Its soo cool and also good    0.6369  High Pos
4      gua selalu nonton wibu , uh asik banget tegang...    0.0000       Neu
...                                                  ...       ...       ...
19995                                              Super    0.5994  High Pos
19996                                               Good    0.4404       Pos
19997           This is amazing I love that application.    0.8402  High Pos
19998                                               good    0.4404       Pos
19999                                               زربة    0.0000       Neu

[20000 rows x 3 columns]

In [22]:
file_name = 'Compound Scores Dataset.xlsx'
review_df.to_excel(file_name)

In [23]:
review_df.value_counts

<bound method DataFrame.value_counts of                                                  Reviews  Compound  Category
0      Updated 3 days before but badly unable to send...   -0.9528  High Neg
1      It's not opening, like literally entered my nu...   -0.2047       Neg
2                                           just awesome    0.6249  High Pos
3                             Its soo cool and also good    0.6369  High Pos
4      gua selalu nonton wibu , uh asik banget tegang...    0.0000       Neu
...                                                  ...       ...       ...
19995                                              Super    0.5994  High Pos
19996                                               Good    0.4404       Pos
19997           This is amazing I love that application.    0.8402  High Pos
19998                                               good    0.4404       Pos
19999                                               زربة    0.0000       Neu

[20000 rows x 3 columns]>

## Building Machine Learning Classification Models based on our Compound Scores.

In [24]:
# STEP 1: Define x and y.

x=review_df['Reviews']
y=review_df['Category']

# STEP 2: Splitting the data into train and test data.

from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.20,random_state=42)

len(x_train), len(x_test), len(y_train), len(y_test)

(16000, 4000, 16000, 4000)

To build a Machine Learning Model, we need to have x as a numerical column. In this case, x is purely a text data and thus we must convert this text into numbers by creating a DTM. You may use a Count DTM or a TF-IDF DTM.

In this case, we will use a TF-IDF DTM as it gives good results.

Let's build a DTM on x_train before building the model.

In [25]:
# let's create a TF-IDF vectorizer.

from sklearn.feature_extraction.text import TfidfVectorizer

vec=TfidfVectorizer()

In [26]:
# Let's create the DTM on x_train.

x_train_dtm=vec.fit_transform(x_train)
x_train_dtm

<16000x9945 sparse matrix of type '<class 'numpy.float64'>'
	with 91669 stored elements in Compressed Sparse Row format>

In [27]:
y_train.value_counts()

Pos         5528
High Pos    5256
Neu         3996
High Neg     614
Neg          606
Name: Category, dtype: int64

In [28]:
# Since data is imbalanced we will use SMOTE to balance the data. 

from imblearn.over_sampling import SMOTE

sm=SMOTE(random_state=42)
x_train_dtm_new,y_train_new=sm.fit_resample(x_train_dtm,y_train)

In [29]:
y_train_new.value_counts()

High Pos    5528
Pos         5528
Neu         5528
Neg         5528
High Neg    5528
Name: Category, dtype: int64

Data is balanced to run predictions. 

## Building Decision Tree Model.

In [30]:
# STEP 1: Create an object.

tree=DecisionTreeClassifier()

# STEP 2: Fit the model object on x and y.

model_dt=tree.fit(x_train_dtm_new,y_train_new)

model_dt

DecisionTreeClassifier()

In [31]:
x_test

10650    Excellent ⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐⭐...
2041                                                    Op
8668                                            Walai,✌️💓💓
1114                  Wow nice app for messaging very nice
13902                                       Woow excellent
                               ...                        
4073                                               Amazing
7442                                                   💖💖💖
9999                                               2months
1870     عالیه فقط میشه یه کاری کنین در صورت دردسترس نب...
15196                                              So nice
Name: Reviews, Length: 4000, dtype: object

In [32]:
# Let's create the x_test_dtm.

x_test_dtm=vec.transform(x_test)
x_test_dtm

<4000x9945 sparse matrix of type '<class 'numpy.float64'>'
	with 21103 stored elements in Compressed Sparse Row format>

In [33]:
# Let's predict the category on x_test_dtm.

y_test_pred_dt=model_dt.predict(x_test_dtm)

y_test_pred_dt

array(['High Pos', 'Neu', 'Neu', ..., 'Neu', 'Neu', 'Pos'], dtype=object)

In [34]:
# Let's find the accuracy of the model by creating a confusion matrix between y_test and y_test_pred.

pd.crosstab(y_test,y_test_pred_dt)

col_0     High Neg  High Pos  Neg  Neu   Pos
Category                                    
High Neg        88        13   29   29     6
Neg             16        22   59   43    15
Neu             17        36   39  897    49
Pos             16        79   26   81  1150
High Pos        23       997   26  132   112

In [35]:
y_test.shape

(4000,)

In [59]:
# The accuracy of model matching both rows and column scores.

Acc_dt=(88+59+897+1150+997)/4000

Acc_dt

0.79775

Accuracy of Decision Tree Model = 79.78%.

## Building Logistic Regression Model.

In [37]:
# STEP 1: Create an object.

lr=LogisticRegression()

# STEP 2: Fit the model object on x and y.

model_lr=lr.fit(x_train_dtm_new,y_train_new)

model_lr

C:\Users\karth\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [38]:
# Let's predict the category on x_test_dtm.

y_test_pred_lr=model_lr.predict(x_test_dtm)

y_test_pred_lr

array(['High Pos', 'Neu', 'Neu', ..., 'Neu', 'Neu', 'Pos'], dtype=object)

In [39]:
# Let's find the accuracy of the model by creating a confusion matrix between y_test and y_test_pred.

pd.crosstab(y_test,y_test_pred_lr)

col_0     High Neg  High Pos  Neg  Neu   Pos
Category                                    
High Neg       108         9   32   15     1
Neg             36         7   87   24     1
Neu             61        25   70  869    13
Pos             30        89   53   71  1109
High Pos        19      1051   17  109    94

In [60]:
# The accuracy of model matching both rows and column scores.

Acc_lr=(108+87+869+1109+1051)/4000

Acc_lr

0.806

Accuracy of Logistic Regression Model = 80.60%.

## Building Random Forest Model.

In [41]:
# STEP 1: Create an object.

rf=RandomForestClassifier()

# STEP 2: Fit the model object on x and y.

model_rf=rf.fit(x_train_dtm_new,y_train_new)

model_rf

RandomForestClassifier()

In [42]:
# Let's predict the category on x_test_dtm.

y_test_pred_rf=model_rf.predict(x_test_dtm)

y_test_pred_rf

array(['High Pos', 'Neu', 'Neu', ..., 'Neu', 'Neu', 'Pos'], dtype=object)

In [43]:
# Let's find the accuracy of the model by creating a confusion matrix between y_test and y_test_pred.

pd.crosstab(y_test,y_test_pred_rf)

col_0     High Neg  High Pos  Neg  Neu   Pos
Category                                    
High Neg        87        16   25   36     1
Neg             10        15   72   54     4
Neu             11        61   17  931    18
Pos              9        96   22  111  1114
High Pos         5      1091   20  125    49

In [61]:
# The accuracy of model matching both rows and column scores.

Acc_rf=(87+72+931+1114+1091)/4000

Acc_rf

0.82375

Accuracy of Random Forest Model = 82.38%.

## Building KNN Model.

In [45]:
# STEP 1: Create an object.

knn=KNeighborsClassifier(n_neighbors=3)

# STEP 2: Fit the model object on x and y.

model_knn=knn.fit(x_train_dtm_new,y_train_new)

model_knn

KNeighborsClassifier(n_neighbors=3)

In [46]:
# Let's predict the category on x_test_dtm.

y_test_pred_knn=model_knn.predict(x_test_dtm)

y_test_pred_knn

array(['High Pos', 'Neu', 'High Pos', ..., 'High Pos', 'High Pos', 'Pos'],
      dtype=object)

In [47]:
# Let's find the accuracy of the model by creating a confusion matrix between y_test and y_test_pred.

pd.crosstab(y_test,y_test_pred_knn)

col_0     High Neg  High Pos  Neg  Neu   Pos
Category                                    
High Neg       102         7   46   10     0
Neg             51         9   78   15     2
Neu            140       341  165  371    21
Pos             80        65  107   48  1052
High Pos       137       828  173   67    85

In [62]:
# The accuracy of model matching both rows and column scores.

Acc_knn=(102+78+371+1052+828)/4000

Acc_knn

0.60775

Accuracy of KNN Model = 60.78%.

## Building XGBoost Model.

In [49]:
# STEP 1: Create an object.

xgb=xgboost.XGBClassifier()

# STEP 2: Fit the model object on x and y.

model_xgb=xgb.fit(x_train_dtm_new,y_train_new)

model_xgb

C:\Users\karth\anaconda3\lib\site-packages\xgboost\sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[19:16:15] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.300000012, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=100, n_jobs=12, num_parallel_tree=1,
              objective='multi:softprob', random_state=0, reg_alpha=0,
              reg_lambda=1, scale_pos_weight=None, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [50]:
# Let's predict the category on x_test_dtm.

y_test_pred_xgb=model_xgb.predict(x_test_dtm)

y_test_pred_xgb

array(['High Pos', 'Neu', 'Neu', ..., 'Neu', 'Neu', 'Pos'], dtype=object)

In [51]:
# Let's find the accuracy of the model by creating a confusion matrix between y_test and y_test_pred.

pd.crosstab(y_test,y_test_pred_xgb)

col_0     High Neg  High Pos  Neg  Neu   Pos
Category                                    
High Neg        96         4   41   23     1
Neg             12         6   86   44     7
Neu             16         9   26  971    16
Pos              9        62   34   99  1148
High Pos         6      1071   21  115    77

In [63]:
# The accuracy of model matching both rows and column scores.

Acc_xgb=(96+86+971+1148+1071)/4000

Acc_xgb

0.843

Accuracy of XGBoost Model = 84.30%.

## From the above 5 classification models, we notice the XGBoost Classification model has highest accuracy of predictions, i.e, 84.3%. Hence we will use the XGBoost Model for predictions of future Review Categories(High Neg, Neg, Neu, Pos, High Pos) based on the DTM of Compound Scores. This will help us suggest changes to the App Developer based on the sentiments of User Reviews extracted and analyzed.

In [64]:
# Let's create a validation data to check if the model is good.

val_data=pd.DataFrame({'Reviews':['Okayish app.']})

val_data

Reviews
0  Okayish app.

In [65]:
# Let's create the val_data into a DTM.

val_data_dtm=vec.transform(val_data)

val_data_dtm

<1x9945 sparse matrix of type '<class 'numpy.float64'>'
	with 1 stored elements in Compressed Sparse Row format>

In [66]:
model_rf.predict(val_data_dtm)

array(['Neu'], dtype=object)

## Validation process for input of Review is giving a 'Neutral' score. Hence, we can infer that our model is running perfectly, according to our set objectives, and this model can be used  for prediction of Compound Scores category of reviews by users in the future, for new reviews.